CMakeLists.txt

```diff
@@ -267,12 +261,18 @@ if(AMENT_ENABLE_TESTING)
   ament_lint_auto_find_test_dependencies()
 endif()
 
-  file(
-    COPY "${PROJECT_NAME}"
-    DESTINATION "${CMAKE_CURRENT_BINARY_DIR}"
-  )
-
-ament_python_install_package(${PROJECT_NAME} PACKAGE_DIR "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}")
+# Copy the Python code to the folder where the C extensions will be generated.
+# This creates a folder in ${CMAKE_CURRENT_BINARY_DIR}/rclpy which has all the
+# necessary Python code and C Python extensions for running tests.
+# This folder is also ready to be installed directly as a Python package.
+file(
+  COPY "${PROJECT_NAME}"
+  DESTINATION "${CMAKE_CURRENT_BINARY_DIR}"
+)
```

Comment:
> Copying files like this should be avoided at all cost.

Why?  That seems to be an absolute assertion without an obvious reason (as if it should be clear to everyone that this is always implicitly true).

> Can someone argue why this is absolutely necessary?

It is necessary because in order to import `rclpy` you need a folder which contains both the Python code (which starts in the source space) and the generated C extensions (`.so`'s) which are generated into the build space. So this copies the Python source into the build space so that there can be a folder in the build space which contains both the Python source and the C extensions. This is needed so that the tests can import `rclpy` as if it were installed, and the tests do so by importing `rclpy` from the build space.

I'd suggest you look into why it is that way by trying to change it (as I did), if you don't believe me. Maybe you can come up with an improvement.